eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs

```diff
@@ -140,17 +163,44 @@ instance
 conwayCertsTransition ::
   forall era.
   ( EraTx era
-  , ShelleyEraTxBody era
+  , ConwayEraTxBody era
+  , ConwayEraPParams era
   , State (EraRule "CERT" era) ~ CertState era
   , Embed (EraRule "CERT" era) (ConwayCERTS era)
-  , Environment (EraRule "CERT" era) ~ DelplEnv era
+  , Environment (EraRule "CERT" era) ~ CertEnv era
   , Signal (EraRule "CERT" era) ~ TxCert era
   ) =>
   TransitionRule (ConwayCERTS era)
 conwayCertsTransition = do
-  TRC (env@(DelegsEnv slot txIx pp tx acnt), certState, certificates) <- judgmentContext
+  TRC
+    ( env@(CertsEnv txIx tx acnt pp slot currentEpoch)
+      , certStateFresh
+      , certificates
+      ) <-
+    judgmentContext
   network <- liftSTS $ asks networkId
 
+  -- Update DRep expiry for all DReps that are voting in this transaction
+  let drepActivity = pp ^. ppDRepActivityL
+      votingDRepCreds =
+        Map.foldlWithKey'
+          ( \acc voter _ -> case voter of
+              DRepVoter cred -> Set.insert cred acc
+              _ -> acc
+          )
+          Set.empty
+          $ unVotingProcedures
+          $ tx ^. bodyTxL . votingProceduresTxBodyL
+      certState =
+        certStateFresh
+          & certVStateL . vsDRepsL
+            %~ Map.mapWithKey
```

Comment:
This is pretty inefficient since you'd have to traverse all of the DReps every time someone submits a transaction, I'd rather build a map with all the credentials that voted this time and then union it with the old DReps map

Reply:
:100: It would be better to iterate over `votingDRepCreds` Set instead and update individual DReps state that way.

Reply:
This logic could even go into the above `foldlWithKey'`. We'd even avoid constructing an extra `votingDRepCreds` set all together

Reply:
:metal:  Great idea! And thanks for the correction! :metal: